drivers/rtc/rtc_stm32.c

```diff
@@ -0,0 +1,322 @@
+}
+
+static void rtc_stm32_disable(struct device *dev)
+{
+	__HAL_RCC_RTC_DISABLE();
+}
+
+
+static int rtc_stm32_set_config(struct device *dev, struct rtc_config *cfg)
+{
+	struct rtc_stm32_data *context = dev->driver_data;
+	RTC_TimeTypeDef sTime;
+	RTC_DateTypeDef sDate;
+	int result = 0;
+
+	if (cfg->cb_fn != NULL) {
+		context->cb_fn = cfg->cb_fn;
+	}
+	sDate.Year = 0;
+	sDate.Month = 0;
+	sDate.Date = 0;
+	sDate.WeekDay = 0;
+
+	if (IS_ENABLED(CONFIG_RTC_STM32_API_REENTRANCY)) {
+		k_sem_take(RP_GET(dev), K_FOREVER);
+	}
+
+	result = HAL_RTC_SetDate(&context->hrtc, &sDate, RTC_FORMAT_BIN);
+	if (result) {
+		result = -EIO;
+	}
+	sTime.Hours      = ((cfg->init_val/1000)/3600)%24;
+	sTime.Minutes    = ((cfg->init_val/1000)%3600)/60;
+	sTime.Seconds    = (cfg->init_val/1000)%60;
+
+	result = HAL_RTC_SetTime(&context->hrtc, &sTime, RTC_FORMAT_BIN);
+	if (result) {
+		result = -EIO;
+	}
+
+	if (IS_ENABLED(CONFIG_RTC_STM32_API_REENTRANCY)) {
+		k_sem_give(RP_GET(dev));
+	}
+
+	return result;
+}
+
+static int rtc_stm32_set_alarm(struct device *dev, const u32_t alarm_val)
+{
+	struct rtc_stm32_data *context = dev->driver_data;
+	RTC_AlarmTypeDef RTC_AlarmStructure;
+	int result = 0;
+
+	/* Set the alarm X+alarm_val */
+	RTC_AlarmStructure.Alarm                = RTC_ALARM_A;
+	RTC_AlarmStructure.AlarmTime.TimeFormat = RTC_HOURFORMAT_24;
+	RTC_AlarmStructure.AlarmTime.Hours      = ((alarm_val/1000)/3600)%24;
+	RTC_AlarmStructure.AlarmTime.Minutes    = ((alarm_val/1000)%3600)/60;
+	RTC_AlarmStructure.AlarmTime.Seconds    = (alarm_val/1000)%60;
+	RTC_AlarmStructure.AlarmDateWeekDay     = ((alarm_val/1000)/3600)/24;
+	RTC_AlarmStructure.AlarmDateWeekDaySel  = RTC_ALARMDATEWEEKDAYSEL_DATE;
+	RTC_AlarmStructure.AlarmMask
+		= RTC_ALARMMASK_DATEWEEKDAY|RTC_ALARMMASK_HOURS;
+	RTC_AlarmStructure.AlarmSubSecondMask   = RTC_ALARMSUBSECONDMASK_NONE;
+	result = HAL_RTC_SetAlarm_IT(&context->hrtc,
+				     &RTC_AlarmStructure,
+				     RTC_FORMAT_BIN);
+	if (result) {
+		result = -EIO;
+	}
+	return result;
+}
+
+static u32_t rtc_stm32_read(struct device *dev)
+{
+	struct rtc_stm32_data *context = dev->driver_data;
+	RTC_TimeTypeDef sTime;
+	RTC_DateTypeDef sDate;
+	u32_t TimeMs;
+	int result;
+
+	result = HAL_RTC_GetDate(&context->hrtc, &sDate, RTC_FORMAT_BIN);
+	if (result) {
+		return -EIO;
+	}
+	result = HAL_RTC_GetTime(&context->hrtc, &sTime, RTC_FORMAT_BIN);
+	if (result) {
+		return -EIO;
+	}
+
+	TimeMs = (60*1000*sTime.Minutes)+(1000*sTime.Seconds) +
+		(1000*sTime.SubSeconds/(sTime.SecondFraction+1));
+
+	return TimeMs;
+}
+
+static u32_t rtc_stm32_get_pending_int(struct device *dev)
+{
+	/* Not available for STM32 RTC */
+	return 0;
+}
+#ifdef CONFIG_SOC_SERIES_STM32F4X
+void rtc_stm32_isr(int line, void *userdata)
+{
+	struct device * const dev = (struct device *)userdata;
+	struct rtc_stm32_data *context = dev->driver_data;
+
+	HAL_RTC_AlarmIRQHandler(&context->hrtc);
+	rtc_data.cb_fn(dev);
+}
+#endif
+#ifdef CONFIG_SOC_SERIES_STM32L4X
```

Comment:
I can't see any difference with the 2 isr implementations (or maybe I am missing something), why doing that differentiation then?

Reply:
You are right. I removed the duplicated code.

Reply:
Signatures for NVIC and EXTI handlers should be different.